From 934a887903476ee90ba1927cca8599a205a60be2 Mon Sep 17 00:00:00 2001
From: ctang <ctang@cloudera.com>
Date: Fri, 22 Jul 2016 14:39:32 -0400
Subject: [PATCH 0695/1164] CDH-41516 HIVE-14267: HS2 open_operations metrics
 not decremented when an operation gets timed out
 (Naveen Gangam, reviewed by Chaoyu Tang, Aihua
 Xu)

Change-Id: I4b62ce4b99a87f9c43e661d5014ba0281f5a9415
---
 .../service/cli/operation/OperationManager.java    |   10 ++++++++++
 .../hive/service/cli/session/HiveSessionImpl.java  |    2 +-
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java b/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
index d7a799b..68c62e1 100644
--- a/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
+++ b/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
@@ -198,6 +198,16 @@ private Operation removeOperation(OperationHandle opHandle) {
   private Operation removeTimedOutOperation(OperationHandle operationHandle) {
     Operation operation = handleToOperation.get(operationHandle);
     if (operation != null && operation.isTimedOut(System.currentTimeMillis())) {
+      LOG.info("Operation is timed out,operation=" + operation.getHandle() + ",state=" + operation.getState().toString());
+      Metrics metrics = MetricsFactory.getInstance();
+      if (metrics != null) {
+        try {
+          metrics.decrementCounter(MetricsConstant.OPEN_OPERATIONS);
+        } catch (Exception e) {
+          LOG.warn("Error decrementing open_operations metric, reported values may be incorrect", e);
+        }
+      }
+
       handleToOperation.remove(operationHandle, operation);
       if (operation instanceof SQLOperation) {
         removeSaveSqlOperationDisplay(operationHandle);
diff --git a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
index f83a994..b41889f 100644
--- a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
+++ b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
@@ -701,7 +701,7 @@ private void closeTimedOutOperations(List<Operation> operations) {
         try {
           operation.close();
         } catch (Exception e) {
-          LOG.warn("Exception is thrown closing timed-out operation " + operation.getHandle(), e);
+          LOG.warn("Exception is thrown closing timed-out operation, reported open_operations metrics may be incorrect " + operation.getHandle(), e);
         }
       }
     } finally {
-- 
1.7.9.5

